### PR TITLE
append template search results after categories list

### PR DESCRIPTION
**Pull request by AbdiTolesa**
Fixes https://github.com/Strategy11/formidable-pro/issues/3962
The order of results is changed to Ascending from Descending but I don't believe out that is much important here or even better this way so I kept the solution as simple one liner to avoid extra DOM query but I can keep the order as before if needed.
In addition to that, I have now excluded the default `Ready Made Solution` and `Check all applications` texts from being accounted for when searching.

**BEFORE**:
image
Searching using keyword `Rea`:
image

**NOW**:
image
Searching using keyword `Rea`:
image

### CODE REVIEW

**Comment by Crabcyborg** [comment]
@AbdiTolesa What do you think about trying to make sure that the elements labelled `Ready Made Solution` and `Check all applications` are excluded from the search? Having these underneath is probably fine either way, but I think this is a bit of a false positive too. I don't think it's useful for "Application" to match all of the applications here. They could just visit the Applications page for that.

**Review by stephywells — CHANGES_REQUESTED** [review]
I'm working on fixing the layout for this in 6.0, but noticed this is working for me. This solves the problem:
`categoryList.insertBefore( getReadyMadeSolution( template ), categoryList.lastChild );`

**Comment by AbdiTolesa** [comment]
> @AbdiTolesa What do you think about trying to make sure that the elements labelled `Ready Made Solution` and `Check all applications` are excluded from the search? Having these underneath is probably fine either way, but I think this is a bit of a false positive too. I don't think it's useful for "Application" to match all of the applications here. They could just visit the Applications page for that.
> 
> I think that would limit the number of Application results enough that keeping them at the top would be fine too.

@Crabcyborg Thanks for suggesting that! It has turned the search results to be relevant for keywords included in the element texts. Indeed, appending the ready made solutions result at the bottom might not make much sense anymore for now but if there are chances the number of applications can grow, we better keep this in the bottom.

File: js/formidable_admin.js [hunk 1]
```diff
@@ -9092,7 +9092,12 @@ function frmAdminBuildJS() {
 		}
 
 		for ( i = 0; i < items.length; i++ ) {
-			var innerText = items[i].innerText.toLowerCase();
+			var innerText;
+			if ( items[i].classList.contains( 'frm-searchable-template' ) ) {
+				innerText = items[i].querySelector( 'h3' ).innerText.toLowerCase();
```
[the remark this answers is not in the document]
Comment (AbdiTolesa) [reply js/formidable_admin.js]: @Crabcyborg I have accounted for that now. Thanks!

File: js/formidable_admin.js [hunk 2]
```diff
@@ -9092,7 +9092,17 @@ function frmAdminBuildJS() {
...
 
 		for ( i = 0; i < items.length; i++ ) {
-			var innerText = items[i].innerText.toLowerCase();
+			var innerText;
+			if ( items[i].classList.contains( 'frm-searchable-template' ) ) {
+				const itemFeaturedForm = items[i].querySelector( '.frm-featured-form' );
+				if ( itemFeaturedForm ) {
+					innerText = itemFeaturedForm.innerText.toLowerCase();
+				} else {
+					innerText = items[i].querySelector( 'h3' ).innerText.toLowerCase();
```
[the remark this answers is not in the document]
Comment (AbdiTolesa) [reply js/formidable_admin.js]: @Crabcyborg Thank you for suggesting that, I have updated this to use `frmDom.search.init` now.

**Comment by Crabcyborg** [comment]
@AbdiTolesa The search isn't quite right still. We want to match the form template descriptions too. If I type "cards", I'd like to see gift card order template.

Screen Shot 2023-02-28 at 2 36 21 PM

Right now in this branch I end up with no results for cards.

Screen Shot 2023-02-28 at 2 37 00 PM

**Comment by AbdiTolesa** [comment]
> @AbdiTolesa The search isn't quite right still. We want to match the form template descriptions too. If I type "cards", I'd like to see gift card order template.
> 
> Screen Shot 2023-02-28 at 2 36 21 PM
> 
> Right now in this branch I end up with no results for cards.
> 
> Screen Shot 2023-02-28 at 2 37 00 PM

@Crabcyborg I have just fixed that, here is the result you see with the same search now.
image

**Comment by Crabcyborg** [comment]
@AbdiTolesa that isn't quite like the screenshot from before still. Ideally we'll still only show the matching templates, not the whole category. In that example, we really only want to see Gift card order.

**Comment by AbdiTolesa** [comment]
> @AbdiTolesa that isn't quite like the screenshot from before still. Ideally we'll still only show the matching templates, not the whole category. In that example, we really only want to see Gift card order.

@Crabcyborg I fixed that now. But I couldn't really find a way to implement other missing functionalities like updating the templates count without altering `frmDom.search.init`, basically I had to trigger a custom event `frmAfterSearch` like we are doing for jQuery and hook into that to do additional tasks. I'm not sure if we should change the event name but I chose that to be consistent with the jQuery event naming used.
image

image

**Comment by Crabcyborg** [comment]
@AbdiTolesa There is a search listener callback built into the frmDom search.

It doesn't use `frmAfterSearch` but when you call init, the params (`init: ( input, targetClassName, { handleSearchResult } = {}) => {`) includes this `handleSearchResult` function.

You can see an example of how it's used for applications to toggle on/off the "no results" element:

```
function handleTemplateSearch({ foundSomething, notEmptySearchText }) {
		if ( false === elements.noTemplateSearchResultsPlaceholder ) {
			elements.noTemplateSearchResultsPlaceholder = getNoResultsPlaceholder();
			elements.templatesGrid.appendChild( elements.noTemplateSearchResultsPlaceholder );
		}
		elements.noTemplateSearchResultsPlaceholder.classList.toggle( 'frm_hidden', ! notEmptySearchText || foundSomething );
	}
```

**Comment by AbdiTolesa** [comment]
> Member



> @AbdiTolesa There is a search listener callback built into the frmDom search.
> 
> It doesn't use `frmAfterSearch` but when you call init, the params (`init: ( input, targetClassName, { handleSearchResult } = {}) => {`) includes this `handleSearchResult` function.
> 
> You can see an example of how it's used for applications to toggle on/off the "no results" element:
> 
> ```
> function handleTemplateSearch({ foundSomething, notEmptySearchText }) {
> 	if ( false === elements.noTemplateSearchResultsPlaceholder ) {
> 		elements.noTemplateSearchResultsPlaceholder = getNoResultsPlaceholder();
> 		elements.templatesGrid.appendChild( elements.noTemplateSearchResultsPlaceholder );
> 	}
> 	elements.noTemplateSearchResultsPlaceholder.classList.toggle( 'frm_hidden', ! notEmptySearchText || foundSomething );
> }
> ```

Thank you for pointing me to that, @Crabcyborg! I have avoided adding new event and use the function callback instead.

File: js/formidable_admin.js [hunk 3]
```diff
@@ -8863,6 +8864,39 @@ function frmAdminBuildJS() {
...
+
+	function initSearch( inputID, itemClass ) {
+		const searchInput = document.getElementById( inputID );
+
+		if ( itemClass === 'control-section accordion-section' ) {
+			itemClass = 'frm-selectable frm-searchable-template';
+
+			const handleTemplateSearch = () => {
+				document.querySelectorAll( '.control-section.accordion-section' ).forEach( category => {
+					const found = category.closest( '.control-section.accordion-section' ).querySelector( '.frm-selectable.frm-searchable-template:not(.frm_hidden)' );
```
Comment (Crabcyborg) [inline js/formidable_admin.js:8879]: Is this `category.closest( '.control-section.accordion-section' )` just trying to check for itself? You're already iterating over '.control-section.accordion-section'. Is this not necessary?

Comment (AbdiTolesa) [reply js/formidable_admin.js]: Yeah it was my mistake indeed, I have removed that extra query from the middle and it works as expected.

**Comment by Crabcyborg** [comment]
@AbdiTolesa Are we able to remove the old after search logic here with this udpate?

https://github.com/Strategy11/formidable-forms/blob/a1d24b6fa9261a2a6ebc1af8687c1d4ac2f39e08/js/formidable_admin.js#L8805

**Comment by AbdiTolesa** [comment]
> frmAfterSearch

@Crabcyborg `frmAfterSearch ` is not getting triggered for templates search anymore as we have switched to the new search, or did I miss your point probably?

**Comment by Crabcyborg** [comment]
@AbdiTolesa That's what I'm saying. There's this block of code here, 23 lines long, that listens for `frmAfterSearch` inside of `#frm_new_form_modal #template-search-input`. It calls `setTemplateCount`, etc, does the things we're doing in this update.

I think it's dead code now, right? It should be safe to remove all of that now.

**Comment by AbdiTolesa** [comment]
> @AbdiTolesa That's what I'm saying. There's this block of code here, 23 lines long, that listens for `frmAfterSearch` inside of `#frm_new_form_modal #template-search-input`. It calls `setTemplateCount`, etc, does the things we're doing in this update.
> 
> I think it's dead code now, right? It should be safe to remove all of that now.

@Crabcyborg Oh, sorry for missing that, I though that was used somewhere else as well wrongly, that is indeed dead now.

**Comment by garretlaxton** [comment]
I don't see the ready made solutions applications if the search is empty, I just see sections.
image

Should I see the applications still?

**Comment by Crabcyborg** [comment]
@garretlaxton when the search is empty, the applications are hidden intentionally. They're really just to match a form template search, in case the template you're looking for is really an application instead.

**Review by garretlaxton — APPROVED** [review]
Gotcha, thanks! They do indeed show if I search. This looks good then.

**Review by Crabcyborg — CHANGES_REQUESTED** [review]
@AbdiTolesa I'm seeing one more thing here.

When I search for a template and then remove the search, my "My Templates" section disappears. This only appears to happen if I have no templates.

tL9SmgdfLW

Can we add a check so the "My Templates" section never disappears when there is nothing in the search value?

**Pull request merged**